# Exploring the Business Dataset

### Read the dataset

In [1]:
import numpy as np
import pandas as pd

In [2]:
%%time
business_df = pd.read_json('../data/business.json', lines=True)
business_df.head()

CPU times: user 2.96 s, sys: 654 ms, total: 3.62 s
Wall time: 3.64 s


business_id                     name              address  \
0  6iYb2HFDywm3zjuRg0shjw      Oskar Blues Taproom         921 Pearl St   
1  tCbdrRPZA0oiIYSmHG3J0w  Flying Elephants at PDX  7000 NE Airport Way   
2  bvN78flM8NLprQ1a1y5dRg           The Reclaimory   4720 Hawthorne Ave   
3  oaepsyvc0J17qwi8cfrOWg              Great Clips   2566 Enterprise Rd   
4  PE9uqAjdw0E4-8mjGl3wVA        Crossfit Terminus  1046 Memorial Dr SE   

          city state postal_code   latitude   longitude  stars  review_count  \
0      Boulder    CO       80302  40.017544 -105.283348    4.0            86   
1     Portland    OR       97218  45.588906 -122.593331    4.0           126   
2     Portland    OR       97214  45.511907 -122.613693    4.5            13   
3  Orange City    FL       32763  28.914482  -81.295979    3.0             8   
4      Atlanta    GA       30316  33.747027  -84.353424    4.0            14   

   is_open                                         attributes  \
0        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
1        1  {'RestaurantsTakeOut': 'True', 'RestaurantsAtt...   
2        1  {'BusinessAcceptsCreditCards': 'True', 'Restau...   
3        1  {'RestaurantsPriceRange2': '1', 'BusinessAccep...   
4        1  {'GoodForKids': 'False', 'BusinessParking': '{...   

                                          categories  \
0  Gastropubs, Food, Beer Gardens, Restaurants, B...   
1  Salad, Soup, Sandwiches, Delis, Restaurants, C...   
2  Antiques, Fashion, Used, Vintage & Consignment...   
3                         Beauty & Spas, Hair Salons   
4  Gyms, Active Life, Interval Training Gyms, Fit...   

                                               hours  
0  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
1  {'Monday': '5:0-18:0', 'Tuesday': '5:0-17:0', ...  
2  {'Thursday': '11:0-18:0', 'Friday': '11:0-18:0...  
3                                               None  
4  {'Monday': '16:0-19:0', 'Tuesday': '16:0-19:0'...

In [3]:
business_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 160585 entries, 0 to 160584
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   business_id   160585 non-null  object 
 1   name          160585 non-null  object 
 2   address       160585 non-null  object 
 3   city          160585 non-null  object 
 4   state         160585 non-null  object 
 5   postal_code   160585 non-null  object 
 6   latitude      160585 non-null  float64
 7   longitude     160585 non-null  float64
 8   stars         160585 non-null  float64
 9   review_count  160585 non-null  int64  
 10  is_open       160585 non-null  int64  
 11  attributes    145593 non-null  object 
 12  categories    160470 non-null  object 
 13  hours         133244 non-null  object 
dtypes: float64(3), int64(2), object(9)
memory usage: 17.2+ MB


### Explode the Categories column

In [4]:
business_df = business_df.assign(categories = business_df['categories'].str.split(', ')).explode('categories')
business_df

business_id                 name              address  \
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
...                        ...                  ...                  ...   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   

                     city state postal_code   latitude   longitude  stars  \
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
...                   ...   ...         ...        ...         ...    ...   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   

        review_count  is_open  \
0                 86        1   
0                 86        1   
0                 86        1   
0                 86        1   
0                 86        1   
...              ...      ...   
160584            82        0   
160584            82        0   
160584            82        0   
160584            82        0   
160584            82        0   

                                               attributes       categories  \
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...       Gastropubs   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...             Food   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...     Beer Gardens   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...      Restaurants   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...             Bars   
...                                                   ...              ...   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...      Restaurants   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...   Middle Eastern   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...    Mediterranean   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...  Persian/Iranian   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...            Salad   

                                                    hours  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
...                                                   ...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  

[709083 rows x 14 columns]

Note that the method above will actually create additional rows which are essentially duplicates, except that each row will contain one of the categories that were previously grouped together. We will deal with these duplicate rows later. Let us double check the shape of our dataset - the number of rows should have increased.

In [5]:
business_df.shape

(709083, 14)

As expected, the number of rows has increased from 160,585 to 709,083.

Let us examine the number of different categories in this dataset.

In [6]:
len(business_df['categories'].unique())

1331

There are 1,331 unique business categories in this dataset! Let's look at the top 40 categories.

In [7]:
business_df['categories'].value_counts().head(40)

Restaurants                  50763
Food                         29469
Shopping                     26205
Beauty & Spas                16574
Home Services                16465
Health & Medical             15102
Local Services               12192
Nightlife                    11990
Bars                         10741
Automotive                   10119
Event Planning & Services     9644
Active Life                   9231
Coffee & Tea                  7725
Sandwiches                    7272
Fashion                       6599
American (Traditional)        6541
Hair Salons                   5900
Pizza                         5756
Hotels & Travel               5703
Breakfast & Brunch            5505
Home & Garden                 5499
American (New)                5458
Real Estate                   5361
Arts & Entertainment          5314
Fast Food                     5300
Auto Repair                   5039
Professional Services         4845
Fitness & Instruction         4741
Doctors             

### Filter for food establishments

We now filter for establishments with the category names of restaurant, food, brunch, and breakfast as save those establishments to a new dataframe.

In [8]:
restaurant_df = (business_df[business_df['categories'].str.contains('restaurant|food|brunch|breakfast', case=False) == True]).reset_index(drop=True)
restaurant_df.head()

business_id                     name              address  \
0  6iYb2HFDywm3zjuRg0shjw      Oskar Blues Taproom         921 Pearl St   
1  6iYb2HFDywm3zjuRg0shjw      Oskar Blues Taproom         921 Pearl St   
2  tCbdrRPZA0oiIYSmHG3J0w  Flying Elephants at PDX  7000 NE Airport Way   
3  D4JtQNTI4X3KcbzacDJsMw      Bob Likes Thai Food         3755 Main St   
4  jFYIsSb7r1QeESVUnXPHBw          Boxwood Biscuit        740 S High St   

        city state postal_code   latitude   longitude  stars  review_count  \
0    Boulder    CO       80302  40.017544 -105.283348    4.0            86   
1    Boulder    CO       80302  40.017544 -105.283348    4.0            86   
2   Portland    OR       97218  45.588906 -122.593331    4.0           126   
3  Vancouver    BC         V5V  49.251342 -123.101333    3.5           169   
4   Columbus    OH       43206  39.947007  -82.997471    4.5            11   

   is_open                                         attributes  \
0        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
1        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
2        1  {'RestaurantsTakeOut': 'True', 'RestaurantsAtt...   
3        1  {'GoodForKids': 'True', 'Alcohol': 'u'none'', ...   
4        1                                               None   

           categories                                              hours  
0                Food  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
1         Restaurants  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
2         Restaurants  {'Monday': '5:0-18:0', 'Tuesday': '5:0-17:0', ...  
3         Restaurants  {'Monday': '17:0-21:0', 'Tuesday': '17:0-21:0'...  
4  Breakfast & Brunch     {'Saturday': '8:0-14:0', 'Sunday': '8:0-14:0'}

In [9]:
restaurant_df.shape

(106453, 14)

In [10]:
len(restaurant_df['categories'].unique())

26

There are less unique number of categories in our dataset. Let's see what these are.

In [11]:
restaurant_df['categories'].unique()

array(['Food', 'Restaurants', 'Breakfast & Brunch', 'Food Court',
       'Seafood', 'Food Delivery Services', 'Food Trucks',
       'Specialty Food', 'Ethnic Food', 'Fast Food', 'Food Stands',
       'Do-It-Yourself Food', 'Seafood Markets', 'Bed & Breakfast',
       'Comfort Food', 'Soul Food', 'Food Safety Training', 'Food Tours',
       'Live/Raw Food', 'Restaurant Supplies', 'Imported Food',
       'Pop-Up Restaurants', 'Food Banks', 'Swiss Food',
       'Dance Restaurants', 'Pub Food'], dtype=object)

In [12]:
restaurant_df['categories'].value_counts()

Restaurants               50763
Food                      29469
Breakfast & Brunch         5505
Fast Food                  5300
Specialty Food             4272
Seafood                    2958
Food Trucks                2537
Food Stands                1321
Food Delivery Services      960
Ethnic Food                 859
Comfort Food                721
Soul Food                   323
Bed & Breakfast             265
Seafood Markets             249
Imported Food               247
Food Court                  214
Do-It-Yourself Food         188
Live/Raw Food               118
Food Tours                   67
Pop-Up Restaurants           62
Restaurant Supplies          33
Food Banks                   16
Pub Food                      2
Food Safety Training          2
Dance Restaurants             1
Swiss Food                    1
Name: categories, dtype: int64

There are some categories here which are not related to our analysis. We will drop any rows associated with these categories below.

In [13]:
categories_to_drop = ['Food Safety Training','Food Trucks', 'Food Delivery Services',
                      'Food Stands', 'Food Court','Restaurant Supplies', 'Food Banks',
                      'Seafood Markets','Imported Food']

# Save the rows that are not categorized by the labels in categories_to_drop
restaurant_df = restaurant_df[~restaurant_df['categories'].isin(categories_to_drop)]
restaurant_df.shape

(100874, 14)

As mentioned earlier, the explode() method created duplicate rows for each business with unique category identifiers. Let's drop those duplicates now.

In [14]:
restaurant_df = restaurant_df.drop_duplicates(subset=['business_id'], ignore_index=True)
restaurant_df.head()

business_id                      name              address  \
0  6iYb2HFDywm3zjuRg0shjw       Oskar Blues Taproom         921 Pearl St   
1  tCbdrRPZA0oiIYSmHG3J0w   Flying Elephants at PDX  7000 NE Airport Way   
2  D4JtQNTI4X3KcbzacDJsMw       Bob Likes Thai Food         3755 Main St   
3  jFYIsSb7r1QeESVUnXPHBw           Boxwood Biscuit        740 S High St   
4  rYs_1pNB_RMtn5WQh55QDA  Chautauqua General Store      100 Clematis Dr   

        city state postal_code   latitude   longitude  stars  review_count  \
0    Boulder    CO       80302  40.017544 -105.283348    4.0            86   
1   Portland    OR       97218  45.588906 -122.593331    4.0           126   
2  Vancouver    BC         V5V  49.251342 -123.101333    3.5           169   
3   Columbus    OH       43206  39.947007  -82.997471    4.5            11   
4    Boulder    CO       80302  39.998449 -105.281006    3.5             5   

   is_open                                         attributes  \
0        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
1        1  {'RestaurantsTakeOut': 'True', 'RestaurantsAtt...   
2        1  {'GoodForKids': 'True', 'Alcohol': 'u'none'', ...   
3        1                                               None   
4        1  {'BikeParking': 'True', 'RestaurantsTakeOut': ...   

           categories                                              hours  
0                Food  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
1         Restaurants  {'Monday': '5:0-18:0', 'Tuesday': '5:0-17:0', ...  
2         Restaurants  {'Monday': '17:0-21:0', 'Tuesday': '17:0-21:0'...  
3  Breakfast & Brunch     {'Saturday': '8:0-14:0', 'Sunday': '8:0-14:0'}  
4                Food  {'Monday': '8:0-20:0', 'Tuesday': '8:0-20:0', ...

In [15]:
restaurant_df.shape

(64079, 14)

We now have 64,079 restaurants in this dataframe.

In [16]:
len(restaurant_df['business_id'].unique())

64079

In [17]:
%%time
restaurant_df = pd.concat([restaurant_df, restaurant_df['attributes'].apply(pd.Series)], axis=1).drop('attributes', axis=1)

CPU times: user 17.6 s, sys: 247 ms, total: 17.9 s
Wall time: 18 s


In [18]:
restaurant_df.shape

(64079, 52)

In [19]:
restaurant_df.columns

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'stars', 'review_count', 'is_open',
       'categories', 'hours', 'RestaurantsTableService', 'WiFi', 'BikeParking',
       'BusinessParking', 'BusinessAcceptsCreditCards',
       'RestaurantsReservations', 'WheelchairAccessible', 'Caters',
       'OutdoorSeating', 'RestaurantsGoodForGroups', 'HappyHour',
       'BusinessAcceptsBitcoin', 'RestaurantsPriceRange2', 'Ambience', 'HasTV',
       'Alcohol', 'GoodForMeal', 'DogsAllowed', 'RestaurantsTakeOut',
       'NoiseLevel', 'RestaurantsAttire', 'RestaurantsDelivery', 'GoodForKids',
       'ByAppointmentOnly', 'Music', 'GoodForDancing', 'BestNights', 'BYOB',
       'CoatCheck', 'Smoking', 'DriveThru', 'BYOBCorkage', 'Corkage',
       'RestaurantsCounterService', 'AcceptsInsurance', 'DietaryRestrictions',
       'AgesAllowed', 'Open24Hours', 'HairSpecializesIn'],
      dtype='object')

In [20]:
restaurant_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64079 entries, 0 to 64078
Data columns (total 52 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   business_id                 64079 non-null  object 
 1   name                        64079 non-null  object 
 2   address                     64079 non-null  object 
 3   city                        64079 non-null  object 
 4   state                       64079 non-null  object 
 5   postal_code                 64079 non-null  object 
 6   latitude                    64079 non-null  float64
 7   longitude                   64079 non-null  float64
 8   stars                       64079 non-null  float64
 9   review_count                64079 non-null  int64  
 10  is_open                     64079 non-null  int64  
 11  categories                  64079 non-null  object 
 12  hours                       53647 non-null  object 
 13  RestaurantsTableService     193

In [21]:
column_dictionaries = ['hours', 'BusinessParking', 'Ambience']

In [22]:
%%time

restaurant_df1 = restaurant_df.copy()
restaurant_df1.iloc[:,15:30]

CPU times: user 49.5 ms, sys: 3.16 ms, total: 52.7 ms
Wall time: 50.8 ms


BikeParking                                    BusinessParking  \
0            True  {'garage': False, 'street': True, 'validated':...   
1           False  {'garage': True, 'street': False, 'validated':...   
2            True  {'garage': False, 'street': True, 'validated':...   
3             NaN                                                NaN   
4            True  {'garage': False, 'street': True, 'validated':...   
...           ...                                                ...   
64074        True  {'garage': False, 'street': False, 'validated'...   
64075        True  {'garage': False, 'street': False, 'validated'...   
64076        True  {'garage': False, 'street': False, 'validated'...   
64077         NaN                                                NaN   
64078        True  {'garage': False, 'street': False, 'validated'...   

      BusinessAcceptsCreditCards RestaurantsReservations WheelchairAccessible  \
0                           True                   False                 True   
1                           True                   False                  NaN   
2                            NaN                    True                  NaN   
3                            NaN                     NaN                  NaN   
4                           True                     NaN                 True   
...                          ...                     ...                  ...   
64074                       True                   False                 True   
64075                       True                   False                  NaN   
64076                       True                     NaN                  NaN   
64077                        NaN                     NaN                  NaN   
64078                       True                    True                 True   

      Caters OutdoorSeating RestaurantsGoodForGroups HappyHour  \
0       True           True                     True      True   
1       True          False                    False       NaN   
2      False          False                     True       NaN   
3        NaN            NaN                      NaN       NaN   
4      False            NaN                      NaN       NaN   
...      ...            ...                      ...       ...   
64074   True           True                     True      True   
64075   True           True                     True       NaN   
64076    NaN            NaN                      NaN       NaN   
64077    NaN            NaN                      NaN       NaN   
64078   True           True                     True       NaN   

      BusinessAcceptsBitcoin RestaurantsPriceRange2  \
0                      False                      2   
1                        NaN                      2   
2                        NaN                      2   
3                        NaN                    NaN   
4                      False                    NaN   
...                      ...                    ...   
64074                    NaN                      2   
64075                    NaN                      2   
64076                    NaN                      2   
64077                    NaN                    NaN   
64078                    NaN                      2   

                                                Ambience  HasTV  \
0      {'touristy': False, 'hipster': False, 'romanti...   True   
1      {'romantic': False, 'intimate': False, 'touris...  False   
2      {'romantic': False, 'intimate': False, 'classy...  False   
3                                                    NaN    NaN   
4                                                    NaN    NaN   
...                                                  ...    ...   
64074  {'touristy': False, 'hipster': False, 'romanti...   True   
64075  {'touristy': False, 'hipster': False, 'romanti...   True   
64076                                                NaN    NaN   
64077                                       

### Expand the hours column

In [23]:
restaurant_df1 = pd.concat([restaurant_df1, restaurant_df1['hours'].apply(pd.Series)], axis=1).drop('hours', axis=1)

In [24]:
restaurant_df1.iloc[:,12:]

RestaurantsTableService     WiFi BikeParking  \
0                        True  u'free'        True   
1                         NaN  u'free'       False   
2                         NaN    u'no'        True   
3                         NaN      NaN         NaN   
4                         NaN      NaN        True   
...                       ...      ...         ...   
64074                    True  u'free'        True   
64075                   False    u'no'        True   
64076                     NaN      NaN        True   
64077                     NaN      NaN         NaN   
64078                    True  u'free'        True   

                                         BusinessParking  \
0      {'garage': False, 'street': True, 'validated':...   
1      {'garage': True, 'street': False, 'validated':...   
2      {'garage': False, 'street': True, 'validated':...   
3                                                    NaN   
4      {'garage': False, 'street': True, 'validated':...   
...                                                  ...   
64074  {'garage': False, 'street': False, 'validated'...   
64075  {'garage': False, 'street': False, 'validated'...   
64076  {'garage': False, 'street': False, 'validated'...   
64077                                                NaN   
64078  {'garage': False, 'street': False, 'validated'...   

      BusinessAcceptsCreditCards RestaurantsReservations WheelchairAccessible  \
0                           True                   False                 True   
1                           True                   False                  NaN   
2                            NaN                    True                  NaN   
3                            NaN                     NaN                  NaN   
4                           True                     NaN                 True   
...                          ...                     ...                  ...   
64074                       True                   False                 True   
64075                       True                   False                  NaN   
64076                       True                     NaN                  NaN   
64077                        NaN                     NaN                  NaN   
64078                       True                    True                 True   

      Caters OutdoorSeating RestaurantsGoodForGroups  ... AgesAllowed  \
0       True           True                     True  ...         NaN   
1       True          False                    False  ...         NaN   
2      False          False                     True  ...         NaN   
3        NaN            NaN                      NaN  ...         NaN   
4      False            NaN                      NaN  ...         NaN   
...      ...            ...                      ...  ...         ...   
64074   True           True                     True  ...         NaN   
64075   True           True                     True  ...         NaN   
64076    NaN            NaN                      NaN  ...         NaN   
64077    NaN            NaN                      NaN  ...         NaN   
64078   True           True                     True  ...         NaN   

      Open24Hours HairSpecializesIn     Monday    Tuesday  Wednesday  \
0             NaN               NaN  11:0-23:0  11:0-23:0  11:0-23:0   
1             NaN               NaN   5:0-18:0   5:0-17:0   5:0-18:0   
2             NaN               NaN  17:0-21:0  17:0-21:0  17:0-21:0   
3             NaN               NaN        NaN        NaN        NaN   
4             NaN               NaN   8:0-20:0   8:0-20:0   8:0-20:0   
...           ...               ...        ...        ...        ...   
64074         NaN               NaN    0:0-0:0   9:0-21:0   9:0-21:0   
64075         NaN               NaN    0:0-0:0    0:0-0:0    0:0-0:0   
64076         NaN               NaN  10:0-18:0  10:0-18:0  10:0-18:0   
64077         NaN               NaN        NaN        NaN        N

In [25]:
restaurant_df2 = restaurant_df1.copy()

# Verfiy
restaurant_df2.head()

business_id                      name              address  \
0  6iYb2HFDywm3zjuRg0shjw       Oskar Blues Taproom         921 Pearl St   
1  tCbdrRPZA0oiIYSmHG3J0w   Flying Elephants at PDX  7000 NE Airport Way   
2  D4JtQNTI4X3KcbzacDJsMw       Bob Likes Thai Food         3755 Main St   
3  jFYIsSb7r1QeESVUnXPHBw           Boxwood Biscuit        740 S High St   
4  rYs_1pNB_RMtn5WQh55QDA  Chautauqua General Store      100 Clematis Dr   

        city state postal_code   latitude   longitude  stars  review_count  \
0    Boulder    CO       80302  40.017544 -105.283348    4.0            86   
1   Portland    OR       97218  45.588906 -122.593331    4.0           126   
2  Vancouver    BC         V5V  49.251342 -123.101333    3.5           169   
3   Columbus    OH       43206  39.947007  -82.997471    4.5            11   
4    Boulder    CO       80302  39.998449 -105.281006    3.5             5   

   ...  AgesAllowed Open24Hours HairSpecializesIn     Monday    Tuesday  \
0  ...          NaN         NaN               NaN  11:0-23:0  11:0-23:0   
1  ...          NaN         NaN               NaN   5:0-18:0   5:0-17:0   
2  ...          NaN         NaN               NaN  17:0-21:0  17:0-21:0   
3  ...          NaN         NaN               NaN        NaN        NaN   
4  ...          NaN         NaN               NaN   8:0-20:0   8:0-20:0   

   Wednesday   Thursday     Friday   Saturday     Sunday  
0  11:0-23:0  11:0-23:0  11:0-23:0  11:0-23:0  11:0-23:0  
1   5:0-18:0   5:0-18:0   5:0-18:0   5:0-18:0   5:0-18:0  
2  17:0-21:0  17:0-21:0  17:0-21:0  17:0-21:0  17:0-21:0  
3        NaN        NaN        NaN   8:0-14:0   8:0-14:0  
4   8:0-20:0   8:0-20:0   8:0-20:0   8:0-20:0   8:0-20:0  

[5 rows x 58 columns]

In [26]:
restaurant_df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64079 entries, 0 to 64078
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   business_id                 64079 non-null  object 
 1   name                        64079 non-null  object 
 2   address                     64079 non-null  object 
 3   city                        64079 non-null  object 
 4   state                       64079 non-null  object 
 5   postal_code                 64079 non-null  object 
 6   latitude                    64079 non-null  float64
 7   longitude                   64079 non-null  float64
 8   stars                       64079 non-null  float64
 9   review_count                64079 non-null  int64  
 10  is_open                     64079 non-null  int64  
 11  categories                  64079 non-null  object 
 12  RestaurantsTableService     19311 non-null  object 
 13  WiFi                        431

In [27]:
import json

In [28]:
%%time

nested_columns = ['Ambience', 'BusinessParking']

for col in nested_columns:
    
    for i in range(0,len(restaurant_df2[col])):

        
        if isinstance(restaurant_df2.loc[i,col], str):
            
            string = restaurant_df2.loc[i, col].replace("'", '"').replace("False", "0").replace("None", "0").replace("True", "1")

            dictionary = json.loads(string)

            restaurant_df2.loc[i,col] = [dictionary]


        else:
            pass
            
    restaurant_df2 = pd.concat([restaurant_df2, restaurant_df2[col].apply(pd.Series)], axis=1).drop([col, 0], axis=1)


CPU times: user 6min 32s, sys: 3.31 s, total: 6min 36s
Wall time: 6min 38s


In [29]:
restaurant_df2

business_id                      name              address  \
0      6iYb2HFDywm3zjuRg0shjw       Oskar Blues Taproom         921 Pearl St   
1      tCbdrRPZA0oiIYSmHG3J0w   Flying Elephants at PDX  7000 NE Airport Way   
2      D4JtQNTI4X3KcbzacDJsMw       Bob Likes Thai Food         3755 Main St   
3      jFYIsSb7r1QeESVUnXPHBw           Boxwood Biscuit        740 S High St   
4      rYs_1pNB_RMtn5WQh55QDA  Chautauqua General Store      100 Clematis Dr   
...                       ...                       ...                  ...   
64074  Q78fYV6B6P6GmX07YVgi4g     Maudie’s Hill Country     12506 Shops Pkwy   
64075  uXdQkuEtvLAzfc3MsO-sTQ      Mama's Cocina Latina  1958 Piedmont Rd NE   
64076  bQX-kwVTyZgcdZGEPzce6Q              Webb's Honey   21777 State Rd 520   
64077  GB75wPibj3IjNauaoCxyGA               Havana Cafe       910 NW 14th St   
64078  ngmLL5Y5OT-bYHKU0kKrYA               Zora Grille  1370 E Altamonte Dr   

                    city state postal_code   latitude   longitude  stars  \
0                Boulder    CO       80302  40.017544 -105.283348    4.0   
1               Portland    OR       97218  45.588906 -122.593331    4.0   
2              Vancouver    BC         V5V  49.251342 -123.101333    3.5   
3               Columbus    OH       43206  39.947007  -82.997471    4.5   
4                Boulder    CO       80302  39.998449 -105.281006    3.5   
...                  ...   ...         ...        ...         ...    ...   
64074           Bee Cave    TX       78738  30.304761  -97.937394    3.0   
64075            Atlanta    GA       30324  33.810996  -84.367084    3.0   
64076            Orlando    FL       32833  28.499836  -81.047478    4.5   
64077           Portland    OR       97209  45.529647 -122.685153    4.0   
64078  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   

       review_count  ...  intimate romantic touristy trendy upscale garage  \
0                86  ...       0.0      0.0      0.0    0.0     0.0    0.0   
1               126  ...       0.0      0.0      0.0    0.0     0.0    1.0   
2               169  ...       0.0      0.0      0.0    0.0     0.0    0.0   
3                11  ...       NaN      NaN      NaN    NaN     NaN    NaN   
4                 5  ...       NaN      NaN      NaN    NaN     NaN    0.0   
...             ...  ...       ...      ...      ...    ...     ...    ...   
64074           310  ...       0.0      0.0      0.0    0.0     0.0    0.0   
64075           185  ...       0.0      0.0      0.0    0.0     0.0    0.0   
64076            18  ...       NaN      NaN      NaN    NaN     NaN    0.0   
64077            10  ...       NaN      NaN      NaN    NaN     NaN    NaN   
64078            82  ...       0.0      0.0      0.0    0.0     0.0    0.0   

       lot street valet validated  
0      0.0    1.0   0.0       0.0  
1      0.0    0.0   0.0       0.0  
2      0.0    1.0   0.0       0.0  
3      NaN    NaN   NaN       NaN  
4      0.0    1.0   0.0       0.0  
...    ...    ...   ...       ...  
64074  1.0    0.0   0.0       0.0  
64075  1.0    0.0   0.0       0.0  
64076  1.0    0.0   0.0       0.0  
64077  NaN    NaN   NaN       NaN  
64078  1.0    0.0   0.0       0.0  

[64079 rows x 70 columns]

In [30]:
restaurant_df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64079 entries, 0 to 64078
Data columns (total 70 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   business_id                 64079 non-null  object 
 1   name                        64079 non-null  object 
 2   address                     64079 non-null  object 
 3   city                        64079 non-null  object 
 4   state                       64079 non-null  object 
 5   postal_code                 64079 non-null  object 
 6   latitude                    64079 non-null  float64
 7   longitude                   64079 non-null  float64
 8   stars                       64079 non-null  float64
 9   review_count                64079 non-null  int64  
 10  is_open                     64079 non-null  int64  
 11  categories                  64079 non-null  object 
 12  RestaurantsTableService     19311 non-null  object 
 13  WiFi                        431

In [31]:
missing_value_columns = (restaurant_df2.isna().sum()/restaurant_df2.shape[0]*100).sort_values(ascending=False)
missing_value_columns

HairSpecializesIn            99.992197
AgesAllowed                  99.968789
Open24Hours                  99.950062
RestaurantsCounterService    99.937577
DietaryRestrictions          99.893881
                               ...    
postal_code                   0.000000
state                         0.000000
city                          0.000000
address                       0.000000
business_id                   0.000000
Length: 70, dtype: float64

In [32]:
restaurant_df2.to_csv('../data/restaurants.csv')